File: frontend/lib/types/crm.ts
// CRM — empresas, contatos e pipeline lidos do Postgres (fora do Notion).

export interface ContatoOut {
  id: string;
  nome: string;
  cargo?: string | null;
  decisor: boolean;
  email?: string | null;
  telefone?: string | null;
  whatsapp?: string | null;
  linkedin?: string | null;
  origem_contato?: string | null;
}

export interface SocioOut {
  id: string;
  nome: string;
  qualificacao?: string | null;
}

export interface EmpresaListItem {
  id: string;
  nome: string;
  cnpj?: string | null;
  site?: string | null;
  cidade?: string | null;
  estado?: string | null;
  setor?: string | null;
  tamanho?: string | null;
  status?: string | null;
  como_conheceu?: string | null;
  score?: number | null;
  n_contatos: number;
}

export interface EmpresaDetalhe {
  id: string;
  nome: string;
  razao_social?: string | null;
  cnpj?: string | null;
  cidade?: string | null;
  estado?: string | null;
  local?: string | null;
  site?: string | null;
  instagram?: string | null;
  facebook?: string | null;
  capital_social?: number | null;
  setor?: string | null;
  tamanho?: string | null;
  score?: number | null;
  status?: string | null;
  como_conheceu?: string | null;
  notas?: string | null;
  analise_json?: Record<string, unknown> | null;
  notion_page_id?: string | null;
  contatos: ContatoOut[];
  socios: SocioOut[];
}

export interface EmpresaListResponse {
  items: EmpresaListItem[];
  total: number;
}

export interface KanbanColuna {
  status: string;
  total: number;
  empresas: EmpresaListItem[];
}

export interface KanbanResponse {
  colunas: KanbanColuna[];
}

export interface CrmMetricas {
  total_empresas: number;
  total_contatos: number;
  total_decisores: number;
  por_status: Record<string, number>;
}

export interface EmpresaRelacionados {
  negocios: NegocioListItem[];
  projetos: ProjetoListItem[];
  atividades: AtividadeListItem[];
}

export type RecordTipo =
  | 'empresa'
  | 'contato'
  | 'negocio'
  | 'projeto'
  | 'atividade';

export interface RecordCampo {
  label: string;
  valor: string;
  campo?: string | null;
  kind?: 'text' | 'num' | 'date' | 'select' | 'bool';
  opcoes_key?: string | null;
  raw?: string | null;
}

export interface RecordLink {
  tipo: RecordTipo;
  id: string;
  nome: string;
  sub?: string | null;
}

export interface RecordGrupo {
  titulo: string;
  itens: RecordLink[];
}

export interface RecordDetalhe {
  tipo: RecordTipo;
  id: string;
  titulo: string;
  campos: RecordCampo[];
  grupos: RecordGrupo[];
  notas?: string | null;
}

export interface EstagioResumo {
  estagio: string;
  total: number;
  valor: number;
}

export interface CrmDashboard {
  pipeline_valor: number;
  pipeline_ponderado: number;
  negocios_abertos: number;
  por_estagio: EstagioResumo[];
  atividades_total: number;
  atividades_pendentes: number;
  atividades_atrasadas: number;
  projetos_total: number;
  projetos_valor_total: number;
  projetos_recebido: number;
  projetos_a_receber: number;
  empresas_total: number;
  clientes_ativos: number;
  contatos_total: number;
}

export interface NotionSyncResultado {
  empresas_lidas: number;
  paginas_ignoradas: number;
  contatos_lidos: number;
  empresas_sem_link: number;
  negocios_lidos: number;
  projetos_lidos: number;
  atividades_lidas: number;
  erros: string[];
}

export interface EmpresaUpsert {
  nome: string;
  razao_social?: string | null;
  cnpj?: string | null;
  site?: string | null;
  cidade?: string | null;
  estado?: string | null;
  local?: string | null;
  instagram?: string | null;
  facebook?: string | null;
  capital_social?: number | null;
  setor?: string | null;
  tamanho?: string | null;
  score?: number | null;
  status?: string | null;
  como_conheceu?: string | null;
  notas?: string | null;
}

export type CrmFacetas = Record<string, string[]>;

export interface CrmOpcao {
  id: string;
  grupo: string;
  valor: string;
  cor?: string | null;
  ordem: number;
  ativo: boolean;
}

export interface MemoriaEvento {
  id: string;
  agente: string;
  alvo_tipo: string;
  alvo_id: string;
  tipo: string;
  resumo?: string | null;
  payload?: Record<string, unknown> | null;
  origem: string;
  created_at: string;
}

export interface MemoriaTimeline {
  alvo_tipo: string;
  alvo_id: string;
  eventos: MemoriaEvento[];
}

export interface OutcomeResumo {
  total: number;
  positivos: number;
  negativos: number;
  taxa_positiva?: number | null;
  por_resultado: Record<string, number>;
  por_alvo_tipo: Record<string, { positivos: number; negativos: number; total: number }>;
}

export interface EmpresasFiltro {
  status?: string;
  setor?: string;
  estado?: string;
  cidade?: string;
  tamanho?: string;
  como_conheceu?: string;
  score_min?: number;
  busca?: string;
  ordenar_por?: string;
  desc?: boolean;
  limit?: number;
}

export interface ContatoListItem {
  id: string;
  nome: string;
  cargo?: string | null;
  decisor: boolean;
  email?: string | null;
  telefone?: string | null;
  whatsapp?: string | null;
  linkedin?: string | null;
  origem_contato?: string | null;
  empresa_id: string;
  empresa_nome?: string | null;
}

export interface ContatoListResponse {
  items: ContatoListItem[];
  total: number;
}

export interface ContatoUpsert {
  empresa_id: string;
  nome: string;
  cargo?: string | null;
  decisor: boolean;
  email?: string | null;
  telefone?: string | null;
  whatsapp?: string | null;
  linkedin?: string | null;
  origem_contato?: string | null;
}

export interface ContatosFiltro {
  busca?: string;
  empresa_id?: string;
  decisor?: boolean;
  origem?: string;
  limit?: number;
}

export interface NegocioListItem {
  id: string;
  nome: string;
  estagio?: string | null;
  valor_estimado?: number | null;
  valor_ponderado?: number | null;
  probabilidade?: string | null;
  origem?: string | null;
  tipo_servico: string[];
  previsao_fechamento?: string | null;
  proxima_acao?: string | null;
  empresa_id?: string | null;
  empresa_nome?: string | null;
  contato_nome?: string | null;
  notas?: string | null;
}

export interface NegocioColuna {
  estagio: string;
  total: number;
  valor_total: number;
  valor_ponderado: number;
  negocios: NegocioListItem[];
}

export interface NegociosPipeline {
  colunas: NegocioColuna[];
  valor_total: number;
  valor_ponderado: number;
}

export interface AtividadeListItem {
  id: string;
  titulo: string;
  tipo?: string | null;
  status?: string | null;
  data?: string | null;
  resumo?: string | null;
  proximos_passos?: string | null;
  negocio_id?: string | null;
  negocio_nome?: string | null;
  contato_nome?: string | null;
}

export interface AtividadeListResponse {
  items: AtividadeListItem[];
  total: number;
}

export interface ProjetoListItem {
  id: string;
  nome: string;
  status?: string | null;
  tipo_servico?: string | null;
  valor_total?: number | null;
  valor_recebido?: number | null;
  a_receber?: number | null;
  prazo_entrega?: string | null;
  data_entrega_real?: string | null;
  link_producao?: string | null;
  repo_github?: string | null;
  empresa_nome?: string | null;
  negocio_nome?: string | null;
  briefing?: string | null;
}

export interface ProjetoListResponse {
  items: ProjetoListItem[];
  total: number;
}

export interface NegocioUpsert {
  nome: string;
  estagio?: string | null;
  valor_estimado?: number | null;
  probabilidade?: string | null;
  origem?: string | null;
  tipo_servico: string[];
  notas?: string | null;
  motivo_perda?: string | null;
  previsao_fechamento?: string | null;
  data_fechamento_real?: string | null;
  proxima_acao?: string | null;
  empresa_id?: string | null;
  contato_id?: string | null;
}

export interface AtividadeUpsert {
  titulo: string;
  tipo?: string | null;
  status?: string | null;
  data?: string | null;
  resumo?: string | null;
  proximos_passos?: string | null;
  negocio_id?: string | null;
  contato_id?: string | null;
}

export interface ProjetoUpsert {
  nome: string;
  status?: string | null;
  tipo_servico?: string | null;
  valor_total?: number | null;
  valor_recebido?: number | null;
  briefing?: string | null;
  link_producao?: string | null;
  repo_github?: string | null;
  forma_pagamento?: string | null;
  prazo_entrega?: string | null;
  data_inicio?: string | null;
  data_entrega_real?: string | null;
  empresa_id?: string | null;
  negocio_id?: string | null;
}

File: backend/app/api/schemas/crm.py
"""Schemas de leitura do CRM (empresas/contatos no Postgres).

O CRM dentro do sistema — substitui abrir o Notion pra consultar. Só leitura
por enquanto; a escrita continua pelo pipeline do Prospector (dual-write).
"""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class ContatoOut(BaseModel):
    id: str
    nome: str
    cargo: str | None = None
    decisor: bool = False
    email: str | None = None
    telefone: str | None = None
    whatsapp: str | None = None
    linkedin: str | None = None
    origem_contato: str | None = None


class SocioOut(BaseModel):
    id: str
    nome: str
    qualificacao: str | None = None


class EmpresaListItem(BaseModel):
    id: str
    nome: str
    cnpj: str | None = None
    site: str | None = None
    cidade: str | None = None
    estado: str | None = None
    setor: str | None = None
    tamanho: str | None = None
    status: str | None = None
    como_conheceu: str | None = None
    score: int | None = None
    n_contatos: int = 0


class EmpresaDetalhe(BaseModel):
    id: str
    nome: str
    razao_social: str | None = None
    cnpj: str | None = None
    cidade: str | None = None
    estado: str | None = None
    local: str | None = None
    site: str | None = None
    instagram: str | None = None
    facebook: str | None = None
    capital_social: float | None = None
    setor: str | None = None
    tamanho: str | None = None
    score: int | None = None
    status: str | None = None
    como_conheceu: str | None = None
    notas: str | None = None
    analise_json: dict | None = None
    notion_page_id: str | None = None
    contatos: list[ContatoOut] = Field(default_factory=list)
    socios: list[SocioOut] = Field(default_factory=list)


class EmpresaListResponse(BaseModel):
    items: list[EmpresaListItem] = Field(default_factory=list)
    total: int = 0


class EmpresaUpsert(BaseModel):
    """Payload de criação/edição de empresa (CRUD do CRM)."""
    nome: str
    razao_social: str | None = None
    cnpj: str | None = None
    site: str | None = None
    cidade: str | None = None
    estado: str | None = None
    local: str | None = None
    instagram: str | None = None
    facebook: str | None = None
    capital_social: float | None = None
    setor: str | None = None
    tamanho: str | None = None
    score: int | None = None
    status: str | None = None
    como_conheceu: str | None = None
    notas: str | None = None


# ── Contatos ─────────────────────────────────────────────────────────

class ContatoListItem(BaseModel):
    id: str
    nome: str
    cargo: str | None = None
    decisor: bool = False
    email: str | None = None
    telefone: str | None = None
    whatsapp: str | None = None
    linkedin: str | None = None
    origem_contato: str | None = None
    empresa_id: str
    empresa_nome: str | None = None


class ContatoListResponse(BaseModel):
    items: list[ContatoListItem] = Field(default_factory=list)
    total: int = 0


class ContatoUpsert(BaseModel):
    """Payload de criação/edição de contato."""
    empresa_id: str
    nome: str
    cargo: str | None = None
    decisor: bool = False
    email: str | None = None
    telefone: str | None = None
    whatsapp: str | None = None
    linkedin: str | None = None
    origem_contato: str | None = None


# ── Negócios (pipeline de vendas) ────────────────────────────────────

class NegocioListItem(BaseModel):
    id: str
    nome: str
    estagio: str | None = None
    valor_estimado: float | None = None
    valor_ponderado: float | None = None   # valor × probabilidade
    probabilidade: str | None = None
    origem: str | None = None
    tipo_servico: list[str] = Field(default_factory=list)
    previsao_fechamento: str | None = None
    proxima_acao: str | None = None
    empresa_id: str | None = None
    empresa_nome: str | None = None
    contato_nome: str | None = None
    notas: str | None = None


class NegocioColuna(BaseModel):
    estagio: str
    total: int
    valor_total: float
    valor_ponderado: float
    negocios: list[NegocioListItem] = Field(default_factory=list)


class NegociosPipeline(BaseModel):
    colunas: list[NegocioColuna] = Field(default_factory=list)
    valor_total: float = 0
    valor_ponderado: float = 0


# ── Atividades ───────────────────────────────────────────────────────

class AtividadeListItem(BaseModel):
    id: str
    titulo: str
    tipo: str | None = None
    status: str | None = None
    data: str | None = None
    resumo: str | None = None
    proximos_passos: str | None = None
    negocio_id: str | None = None
    negocio_nome: str | None = None
    contato_nome: str | None = None


class AtividadeListResponse(BaseModel):
    items: list[AtividadeListItem] = Field(default_factory=list)
    total: int = 0


# ── Projetos ─────────────────────────────────────────────────────────

class ProjetoListItem(BaseModel):
    id: str
    nome: str
    status: str | None = None
    tipo_servico: str | None = None
    valor_total: float | None = None
    valor_recebido: float | None = None
    a_receber: float | None = None
    prazo_entrega: str | None = None
    data_entrega_real: str | None = None
    link_producao: str | None = None
    repo_github: str | None = None
    empresa_nome: str | None = None
    negocio_nome: str | None = None
    briefing: str | None = None


class ProjetoListResponse(BaseModel):
    items: list[ProjetoListItem] = Field(default_factory=list)
    total: int = 0


class EmpresaRelacionados(BaseModel):
    """Tudo ligado a uma empresa — a ficha 360."""
    negocios: list[NegocioListItem] = Field(default_factory=list)
    projetos: list[ProjetoListItem] = Field(default_factory=list)
    atividades: list[AtividadeListItem] = Field(default_factory=list)


# ── Record universal (navegação relacional bidirecional) ─────────────

class RecordCampo(BaseModel):
    label: str
    valor: str                      # texto exibido (formatado)
    campo: str | None = None        # nome do campo no banco (None = não editável)
    kind: str = "text"              # text|num|date|select|bool
    opcoes_key: str | None = None   # chave em /crm/opcoes (quando kind=select)
    raw: str | None = None          # valor cru pra edição (default = valor)


class RecordLink(BaseModel):
    tipo: str           # empresa|contato|negocio|projeto|atividade
    id: str
    nome: str
    sub: str | None = None


class RecordGrupo(BaseModel):
    titulo: str
    itens: list[RecordLink] = Field(default_factory=list)


class RecordDetalhe(BaseModel):
    tipo: str
    id: str
    titulo: str
    campos: list[RecordCampo] = Field(default_factory=list)
    grupos: list[RecordGrupo] = Field(default_factory=list)
    notas: str | None = None


class RecordPatch(BaseModel):
    """Edição parcial: só os campos enviados são alterados (edição inline)."""
    campos: dict[str, Any]


# ── Opções de select gerenciáveis ────────────────────────────────────

class OpcaoOut(BaseModel):
    id: str
    grupo: str
    valor: str
    cor: str | None = None
    ordem: int
    ativo: bool


class OpcaoCreate(BaseModel):
    grupo: str
    valor: str
    cor: str | None = None


class OpcaoUpdate(BaseModel):
    valor: str | None = None
    cor: str | None = None
    ativo: bool | None = None


class OpcaoReorder(BaseModel):
    grupo: str
    ids: list[str] = Field(default_factory=list)


# ── Upserts (CRUD) ───────────────────────────────────────────────────

class NegocioUpsert(BaseModel):
    nome: str
    estagio: str | None = None
    valor_estimado: float | None = None
    probabilidade: str | None = None
    origem: str | None = None
    tipo_servico: list[str] = Field(default_factory=list)
    notas: str | None = None
    motivo_perda: str | None = None
    previsao_fechamento: str | None = None      # YYYY-MM-DD
    data_fechamento_real: str | None = None
    proxima_acao: str | None = None
    empresa_id: str | None = None
    contato_id: str | None = None


class AtividadeUpsert(BaseModel):
    titulo: str
    tipo: str | None = None
    status: str | None = None
    data: str | None = None                     # ISO (date ou datetime)
    resumo: str | None = None
    proximos_passos: str | None = None
    negocio_id: str | None = None
    contato_id: str | None = None


class ProjetoUpsert(BaseModel):
    nome: str
    status: str | None = None
    tipo_servico: str | None = None
    valor_total: float | None = None
    valor_recebido: float | None = None
    briefing: str | None = None
    link_producao: str | None = None
    repo_github: str | None = None
    forma_pagamento: str | None = None
    prazo_entrega: str | None = None
    data_inicio: str | None = None
    data_entrega_real: str | None = None
    empresa_id: str | None = None
    negocio_id: str | None = None


class KanbanColuna(BaseModel):
    status: str
    total: int
    empresas: list[EmpresaListItem] = Field(default_factory=list)


class KanbanResponse(BaseModel):
    colunas: list[KanbanColuna] = Field(default_factory=list)


class CrmMetricas(BaseModel):
    total_empresas: int = 0
    total_contatos: int = 0
    total_decisores: int = 0
    por_status: dict[str, int] = Field(default_factory=dict)


class EstagioResumo(BaseModel):
    estagio: str
    total: int
    valor: float


class CrmDashboard(BaseModel):
    # Pipeline
    pipeline_valor: float = 0
    pipeline_ponderado: float = 0
    negocios_abertos: int = 0
    por_estagio: list[EstagioResumo] = Field(default_factory=list)
    # Atividades
    atividades_total: int = 0
    atividades_pendentes: int = 0
    atividades_atrasadas: int = 0
    # Projetos (entrega/financeiro)
    projetos_total: int = 0
    projetos_valor_total: float = 0
    projetos_recebido: float = 0
    projetos_a_receber: float = 0
    # Contas
    empresas_total: int = 0
    clientes_ativos: int = 0
    contatos_total: int = 0
